fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorModelRegistry.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.item.v1;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+import net.minecraft.client.render.entity.model.BipedEntityModel;
+import net.minecraft.entity.EquipmentSlot;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.item.ItemConvertible;
+import net.minecraft.item.ItemStack;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.impl.client.item.ArmorModelRegistryImpl;
+
+/**
+ * A class for registering custom armor models for {@link ItemConvertible}, to be provided by a {@link Provider}.
+ *
+ * <p>
+ * This can be used to replace existing vanilla armor models conditionally, however each {@link net.minecraft.item.Item}
+ * instance can only allow one {@link Provider}.
+ * </p>
+ */
+@Environment(EnvType.CLIENT)
+public final class ArmorModelRegistry {
+	private ArmorModelRegistry() {
+	}
+
+	/**
+	 * Registers a provider for custom armor models for an item.
+	 *
+	 * @param provider the provider for the model
+	 * @param items    the items to be registered for
+	 */
+	public static void register(/* @Nullable */ Provider provider, ItemConvertible... items) {
```

Comment:
Not sure about using ItemConvertible over Item. Blocks are unusual for armor and for the rare exception the caller can convert easily. Item may be less confusing. The implementation mandates ArmorItem, so maybe we should use/require that in the API as well.

Reply:
Hmm, can't carved pumpkins be used as armor

Reply:
Carved pumpkins aren't armor but `BlockItem`s and they use their item model instead of an armor model (this behavior is probably specific to the head slot).

Reply:
The instrumented code starts with `if (itemStack.getItem() instanceof ArmorItem) {`

Reply:
More precisely, the pumpkin or worn heads use separate feature renderers that aren't covered by this PR.